prov/shm/src/smr_msg.c

```diff
@@ -377,126 +349,68 @@ struct fi_ops_msg smr_msg_ops = {
 	.injectdata = smr_injectdata,
 };
 
-static inline struct smr_ep_entry *smr_get_trecv_entry(struct smr_ep *ep, uint64_t flags)
+ssize_t smr_generic_trecv(struct smr_ep *ep, const struct iovec *iov,
+			  size_t iov_count, fi_addr_t addr, void *context,
+			  uint64_t tag, uint64_t ignore, uint64_t flags)
 {
 	struct smr_ep_entry *entry;
+	ssize_t ret = -FI_EAGAIN;
 
-	if (freestack_isempty(ep->recv_fs))
-		return NULL;
-
-	entry = freestack_pop(ep->recv_fs);
-	entry->err = 0;
-	entry->flags = smr_convert_rx_flags(flags);
-
-	return entry;
-}
+	assert(iov_count <= SMR_IOV_LIMIT);
 
-static inline ssize_t
-smr_proccess_trecv_post(struct smr_ep *ep, struct smr_ep_entry *entry)
-{
-	ssize_t ret;
+	fastlock_acquire(&ep->region->lock);
+	fastlock_acquire(&ep->util_ep.rx_cq->cq_lock);
 
-	ret = smr_progress_unexp(ep, entry, &ep->unexp_tagged_queue);
-	if (!ret || ret == -FI_EAGAIN)
-		return ret;
+	entry = smr_get_recv_entry(ep, iov, iov_count, addr, context, tag,
+				   ignore, flags);
+	if (!entry)
+		goto out;
 
 	dlist_insert_tail(&entry->entry, &ep->trecv_queue.list);
-	return 0;
+	ret = smr_progress_unexp_queue(ep, entry, &ep->unexp_tagged_queue);
```

Comment:
The smr_generic_recv/trecv are basically the same.  You could create a single function if you pass the recv_queue and unexp_queue to it.